blueoil/utils/config.py

```diff
@@ -216,6 +216,10 @@ def _save_config_yaml(output_dir, config):
     config_dict = _easy_dict_to_dict(config)
     file_path = os.path.join(output_dir, file_name)
 
+    # For this particular class name, the parent class is more appropriate for the config.
+    if str(config_dict['DATASET_CLASS']) == "<class 'abc.DATASET_CLASS'>":
```

Comment:
I'm not sure the type of `config_dict['DATASET_CLASS']`, but maybe we can write like
`if config_dict['DATASET_CLASS'] == abc.DATASET_CLASS`

Reply:
yes, it's not ideal to compare using the `str` function, but I'm not sure how to avoid it.
The type of `config_dict['DATASET_CLASS']` is `<class 'abc.ABCMeta'>` In both cases. However, it is necessary to distinguish between the two cases. So, an example for each of the two cases:
1.
using the dataset module [cifar10.py](https://github.com/blue-oil/blueoil/blob/master/blueoil/datasets/cifar10.py)
`type(config_dict['DATASET_CLASS'])` is `<class 'abc.ABCMeta'>`
`str(config_dict['DATASET_CLASS'])` is `<class 'blueoil.datasets.cifar100.Cifar100'>`
`config_dict['DATASET_CLASS'].__bases__[0]` is `<class 'blueoil.datasets.base.Base'>`
2.
using `python blueoil/cmd/main.py init` to generate the dataset class
`type(config_dict['DATASET_CLASS'])` is `<class 'abc.ABCMeta'>`
`str(config_dict['DATASET_CLASS'])` is `<class 'abc.DATASET_CLASS'>`
`config_dict['DATASET_CLASS'].__bases__[0]` is `<class 'blueoil.datasets.image_folder.ImageFolderBase'>`

In the case of 2, the useful thing to save is the base class, but in case 1 we can save in the usual way. If we only see the type, then we can't know which thing to do. But if seeing the string of the object, it becomes clear what to do.

Reply:
Got it. How about using `isinstance` ?

Reply:
ah, right, `isinstance` would be a pythonic way to do it. Thank you, I will try it out

Reply:
I tried it out, but not much luck. I write the main details down here. Since it is a class, I think `issubclass` function makes sense. However, it's difficult to access the class, or a parent of it.

An example line of the config file created by `blueoil/cmd/main.py init` is:
`DATASET_CLASS = type('DATASET_CLASS', (ImageFolderBase,), {'extend_dir': 'cifar/train', 'validation_extend_dir': 'cifar/test'})`
then in line 133 of [utils/config.py](https://github.com/blue-oil/blueoil/blob/383aebfd4a1a4150ef4a09b7b1d1221871a4ead7/blueoil/utils/config.py#L133) (loading the config for starting training), the exec command generates the class `<class 'abc.DATASET_CLASS'>`.

I thought maybe it's possible to do something like `issubclass(config_dict['DATASET_CLASS'], type('DATASET_CLASS', (), {}))`, but I think this will always evaluate False, because python will see this as a definition of a new class.
In fact, even using `type('DATASET_CLASS', (ImageFolderBase,), {'extend_dir': 'cifar/train', 'validation_extend_dir': 'cifar/test'})` and comparing with `issubclass` evaluates as False in the previous example.

Anyway, I think doing this is no better than using `str(config_dict['DATASET_CLASS'])`. The benefit of using `issubclass` would be if we had a module which created the class in the first place. But what we actually have is `blueoil/cmd/main.py init` generates a config containing the definition of the class, and the class is generated from that.

I feel like to avoid using the check `str(config_dict['DATASET_CLASS'])` (or similar), then it is necessary to change the behaviour of how `DATASET_CLASS` is used for generated configs. I think `<class 'abc.DATASET_CLASS'>` is the class of generated configs (including the configs used in the tests).

Reply:
@joelN123 I heard iizukak-san will contact you so could you please re-assign me as a reviewer after the discussion?

Reply:
yes, that's right. ok, I will remember to do that. Thank you for your time on this PR.